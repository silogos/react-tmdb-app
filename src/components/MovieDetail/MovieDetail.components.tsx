import { NavLink } from "react-router";
import { ChevronLeftIcon } from "@heroicons/react/16/solid";
import {
  CalendarDaysIcon,
  ClockIcon,
  StarIcon,
} from "@heroicons/react/16/solid";

import { getReleaseYear } from "@/utils/Movie.utils";

import {
  getRatingText,
  formatRunTime,
  formatReleaseDate,
} from "@/utils/Movie.utils";
import Banner from "../Banner";
import Poster from "../Poster";
import {
  MovieDetailHeaderProps,
  MovieDetailHeroProps,
} from "./MovieDetail.type";

export function MovieDetailHeader({ data }: MovieDetailHeaderProps) {
  return (
    <header className="overflow-hidden mb-2 bg-gray-300 bg-opacity-20">
      <div className="mx-auto max-w-screen-xl px-8 py-4 text-gray-400">
        <div className="flex flex-row gap-4 items-stretch">
          <Poster poster={data.poster_path} className="w-20" />

          <div className="flex flex-col justify-evenly">
            <h1 className="text-left text-white text-xl lg:text-4xl">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>

            <NavLink
              to={".."}
              className="cursor-pointer hover:underline flex flex-row items-center text-base lg:text-lg"
            >
              <ChevronLeftIcon className="size-6 inline" />
              <span>Back to list</span>
            </NavLink>
          </div>
        </div>
      </div>
    </header>
  );
}

export function MovieDetailHero({ data }: MovieDetailHeroProps) {
  return (
    <section id="hero" className="relative overflow-hidden">
      <div className="mx-auto max-w-screen-xl px-8 py-8">
        <div className="flex flex-col lg:flex-row gap-10">
          <Banner className="lg:h-96" src={data.backdrop_path} />

          <div className="py-4">
            <h1 className="text-left text-white text-xl lg:text-4xl mb-4">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              {data.genres.map((genre) => (
                <div
                  key={genre.id}
                  className="border border-gray-300 text-sm px-2 py-1 rounded-full "
                >
                  {genre.name}
                </div>
              ))}
            </div>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              <span className="text-left font-bold">
                <CalendarDaysIcon className="size-5 inline mr-1 mb-1" />
                {formatReleaseDate(data.release_date)}
              </span>
              <span>•</span>
              <span className="text-left font-bold">
                <StarIcon className="size-5 inline mr-1 mb-1" />
                {getRatingText(data.vote_average)}
              </span>
              <span>•</span>
              <span className="text-left font-bold">
                <ClockIcon className="size-5 inline mr-1" />
                {formatRunTime(data.runtime)}
              </span>
            </div>

            <p className="text-left text-gray-300 text-base my-4">
              <i>{data.tagline}</i>
            </p>

            <span className="text-left text-white text-lg font-bold mb-1">
              Overview
            </span>
            <p className="text-left text-gray-300 text-base">{data.overview}</p>
          </div>
        </div>
      </div>
    </section>
  );
}
