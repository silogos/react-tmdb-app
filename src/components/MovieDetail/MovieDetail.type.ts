import { MovieDetail } from "@/types/Movie.types";

export type MovieDetailHeaderProps = {
  data: MovieDetail;
};
export type MovieDetailHeroProps = {
  data: MovieDetail;
};
