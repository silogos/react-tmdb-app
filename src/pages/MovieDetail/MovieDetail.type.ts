import { MovieDetail } from "@/types/Movie.types";

export type MovieDetailHeroProps = {
  data: MovieDetail;
};
