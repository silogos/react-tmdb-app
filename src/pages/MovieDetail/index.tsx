export { default } from "./MovieDetail.page";
