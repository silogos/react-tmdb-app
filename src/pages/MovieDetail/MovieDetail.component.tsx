import { NavLink } from "react-router";
import { ChevronLeftIcon } from "@heroicons/react/16/solid";
import {
  CalendarDaysIcon,
  ClockIcon,
  StarIcon,
} from "@heroicons/react/16/solid";

import {
  getRatingText,
  formatRunTime,
  formatReleaseDate,
  getReleaseYear,
} from "@/utils/Movie.utils";

import { MovieDetailHeroProps } from "./MovieDetail.type";
import Banner from "@/components/Banner";

export function Header() {
  return (
    <header className="overflow-hidden mb-2 bg-gray-300 bg-opacity-20">
      <div className="mx-auto max-w-screen-xl px-8 py-4 text-gray-400">
        <div className="flex flex-row gap-4 items-stretch">
          <div className="flex flex-row justify-evenly gap-2 items-center">
            <NavLink
              to={".."}
              className="cursor-pointer hover:underline flex flex-row items-center text-sm"
            >
              <ChevronLeftIcon className="size-6 inline" />
              Back to list
            </NavLink>
          </div>
        </div>
      </div>
    </header>
  );
}

export function Hero({ data }: MovieDetailHeroProps) {
  return (
    <section id="hero" className="relative overflow-hidden">
      <div className="mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8">
        <div className="flex flex-col lg:flex-row gap-2 lg:gap-10">
          <Banner className="lg:h-96" src={data.backdrop_path} />

          <div className="py-4">
            <h1 className="text-left text-white text-xl sm:text-2xl lg:text-4xl mb-4">
              {data.title} ({getReleaseYear(data.release_date)})
            </h1>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-base">
              {data.genres.map((genre) => (
                <div
                  key={genre.id}
                  className="border border-gray-300 text-xs lg:text-sm px-2 py-1 rounded-full "
                >
                  {genre.name}
                </div>
              ))}
            </div>
            <div className="flex flex-row flex-wrap items-center gap-3 mb-4 text-gray-300 text-xs sm:text-sm lg:text-base">
              <span className="text-left font-bold text-nowrap">
                <CalendarDaysIcon className="size-5 inline mr-1 mb-1" />
                {formatReleaseDate(data.release_date)}
              </span>
              <span>•</span>
              <span className="text-left font-bold text-nowrap">
                <StarIcon className="size-5 inline mr-1 mb-1" />
                {getRatingText(data.vote_average)}
              </span>
              <span>•</span>
              <span className="text-left font-bold text-nowrap">
                <ClockIcon className="size-5 inline mr-1" />
                {formatRunTime(data.runtime)}
              </span>
            </div>

            <p className="text-left text-gray-300 text-xs sm:text-sm lg:text-base my-4">
              <i>{data.tagline}</i>
            </p>

            <p className="text-left text-white text-sm sm:text-base lg:text-lg font-bold mb-1">
              Overview
            </p>
            <p className="text-left text-gray-300 text-xs sm:text-sm lg:text-base">
              {data.overview}
            </p>
          </div>
        </div>
      </div>
    </section>
  );
}

export function HeroShimmering() {
  return (
    <div className="animate-pulse mx-auto max-w-screen-xl px-8 py-2 sm:py-4 lg:py-8">
      <div className="flex flex-col lg:flex-row gap-8 lg:gap-10">
        <div
          className={`relative lg:h-96 aspect-video bg-gray-700 bg-opacity-90 rounded-lg overflow-hidden`}
        ></div>
        <div className="flex-1 flex flex-col gap-4">
          <div className="w-full h-8 lg:h-10 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="flex flex-row gap-4">
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
            <div className="w-20 h-6 lg:h-8 rounded-full bg-gray-700 bg-opacity-90" />
          </div>
          <div className="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-3/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-2/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
          <div className="w-1/4 h-6 lg:h-8 rounded-lg bg-gray-700 bg-opacity-90" />
        </div>
      </div>
    </div>
  );
}
