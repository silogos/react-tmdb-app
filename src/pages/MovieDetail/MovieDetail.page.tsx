import CreditList from "@/components/CreditList";
import useMovieDetail from "@/hooks/useMovieDetail";
import useCredits from "@/hooks/useCredits";

import { Header, Hero, HeroShimmering } from "./MovieDetail.component";
import useBackgroundImage from "@/hooks/useBackgroundImage";
import { DisplayError } from "@/components/Displays";

function MovieDetailPage() {
  const { data, isLoading, isError, getData } = useMovieDetail();
  const { casts, crews } = useCredits();
  useBackgroundImage(data?.backdrop_path);

  return (
    <>
      <div className="relative w-full min-h-screen">
        <Header />

        {isError && (
          <div className="mx-auto max-w-screen-xl p-8">
            <DisplayError onReload={getData} />
          </div>
        )}

        {isLoading && <HeroShimmering />}

        {!isLoading && data && (
          <>
            <Hero data={data} />
            <CreditList title="Cast" credits={casts} />
            <CreditList title="Crew" credits={crews} />
          </>
        )}
      </div>
    </>
  );
}

export default MovieDetailPage;
