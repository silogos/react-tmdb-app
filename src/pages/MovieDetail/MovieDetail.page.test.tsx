import { describe, it, expect } from "vitest";
import MovieDetail from "./MovieDetail.page";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router";

describe("MovieDetail Page", () => {
  it("renders correctly", () => {
    const component = render(
      <BrowserRouter>
        <MovieDetail />
      </BrowserRouter>
    );
    expect(component).toBeTruthy();
  });

  it("renders toMatchSnapshot", () => {
    const component = render(
      <BrowserRouter>
        <MovieDetail />
      </BrowserRouter>
    );
    expect(component).toMatchSnapshot();
  });
});
