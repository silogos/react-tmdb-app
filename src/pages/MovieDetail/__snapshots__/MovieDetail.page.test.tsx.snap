// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieDetail Page > renders data ready toMatchSnapshot 1`] = `
<div>
  <div
    class="relative w-full min-h-screen"
  >
    <div
      data-testid="Header"
    />
    <div
      data-testid="Hero"
    />
    <div
      data-testid="CreditList"
    />
    <div
      data-testid="CreditList"
    />
  </div>
</div>
`;

exports[`MovieDetail Page > renders initial data toMatchSnapshot 1`] = `
<div>
  <div
    class="relative w-full min-h-screen"
  >
    <div
      data-testid="Header"
    />
    <div
      data-testid="HeroShimmering"
    />
  </div>
</div>
`;
